pkg/promtail/targets/pubsub/formatter.go

```diff
@@ -0,0 +1,71 @@
+package pubsub
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/prometheus/common/model"
+)
+
+// LogEntry that will be written to the pubsub topic.
+// According to the following spec.
+// https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+type GCPLogEntry struct {
+	LogName  string `json:"logName"`
+	Resource struct {
+		Type   string            `json:"type"`
+		Labels map[string]string `json:"labels"`
+	} `json:"resource"`
+	Timestamp string `json:"timestamp"`
+
+	// The time the log entry was received by Logging.
+	// Its important that `Timestamp` is optional in GCE log entry.
+	ReceiveTimestamp string `json:"receiveTimestamp"`
+
+	// TODO(kavi): Add other meta data fields as well if needed.
+}
+
+func format(m *pubsub.Message, other model.LabelSet) (api.Entry, error) {
+	var ge GCPLogEntry
```

Comment:
What should we do with pubsub message that are not `GCPLogEntry` ? Shouldn't we also handle them ?

I think if it's a `textPayload` we should use it as log line. see
![image](https://user-images.githubusercontent.com/1053421/103650188-a678f680-4f2d-11eb-9262-0f830428b90a.png)

I'm not sure if all labels should be included but I don't see a good way of dealing with this.

Reply:
>What should we do with pubsub message that are not GCPLogEntry ? Shouldn't we also handle them ?
No. we don't. 
these topic is subscribed only for GCP log entry (all logs from GCP have same format). Other messages we don't process.

>I think if it's a textPayload we should use it as log line
Payload can be one of textPayload, protoPayload or JsonPayload. we have handle all the three.
are you suggesting to put only these payload as log entry? currently I'm treating entire log message as log entry to Loki

Reply:
for textPayload I would expect the log to be the payload yes.

> No. we don't.

Kinda weird that the name of the target is pubsub then should we renamed to GCP Sink or something that sounds closer to the reality ? @slim-bean

Reply:
>for textPayload I would expect the log to be the payload yes.

I didn't understand. then we loose other key values. Is that fine?

example this is the complete log entry we send to loki
```
{
  "insertId": "ajv4d1f1ch8dr",
  "logName": "projects/grafanalabs-dev/logs/cloudaudit.googleapis.com%2Fdata_access",
  "protoPayload": {
    "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
    "authenticationInfo": {
      "principalEmail": "1040409107725-compute@developer.gserviceaccount.com",
      "serviceAccountDelegationInfo": [
        {
          "firstPartyPrincipal": {
            "principalEmail": "service-1040409107725@compute-system.iam.gserviceaccount.com"
          }
        }
      ]
    },
    "authorizationInfo": [
      {
        "granted": true,
        "permission": "storage.objects.list",
        "resource": "projects/_/buckets/dev-us-central1-cortex-tsdb-dev",
        "resourceAttributes": {
        }
      },
      {
        "permission": "storage.objects.get",
        "resource": "projects/_/buckets/dev-us-central1-cortex-tsdb-dev/objects/load-generator-20/01EM34PFBC2SCV3ETBGRAQZ090/deletion-mark.json",
        "resourceAttributes": {
        }
      }
    ],
    "methodName": "storage.objects.get",
    "requestMetadata": {
      "callerIp": "34.66.19.193",
      "callerNetwork": "//compute.googleapis.com/projects/grafanalabs-dev/global/networks/__unknown__",
      "callerSuppliedUserAgent": "thanos-store-gateway/1.5.0 (go1.14.9),gzip(gfe)",
      "destinationAttributes": {
      },
      "requestAttributes": {
        "auth": {
        },
        "time": "2021-01-01T02:17:10.661405637Z"
      }
    },
    "resourceLocation": {
      "currentLocations": [
        "us-central1"
      ]
    },
    "resourceName": "projects/_/buckets/dev-us-central1-cortex-tsdb-dev/objects/load-generator-20/01EM34PFBC2SCV3ETBGRAQZ090/deletion-mark.json",
    "serviceName": "storage.googleapis.com",
    "status": {
    }
  },
  "receiveTimestamp": "2021-01-01T02:17:10.82013623Z",
  "resource": {
    "labels": {
      "bucket_name": "dev-us-central1-cortex-tsdb-dev",
      "location": "us-central1",
      "project_id": "grafanalabs-dev"
    },
    "type": "gcs_bucket"
  },
  "severity": "INFO",
  "timestamp": "2021-01-01T02:17:10.655982344Z"
}
```
in the above example its `protopayload` (you can see the key `protopayload` in the above json). So currently, no matter what we send this complete entry to loki. 

Are you suggesting to send only the value `textPayload` to loki if its type is `textPayload`?? Won't we loose other key values? May I miss something?

>Kinda weird that the name of the target is pubsub then should we renamed to GCP Sink or something that sounds closer to the reality ?

I agree, we can rename the entire target manager to `pubsub-gcp`(or something similar). Because anyway we cannot use  this targetmanager for AWS.. AWS may need whole new set of configs to pull logs out of it.

Is that what you mean? If that's the case. sure I will change it.

Reply:
> Are you suggesting to send only the value textPayload to loki if its type is textPayload?? Won't we loose other key values? May I miss something?

Yes. Still same labels, but the content should be textPayload, other e.g protoPayload are ok to send everything. This is because I believe those are custom application log from the cloud user.

>I agree, we can rename the entire target manager to pubsub-gcp(or something similar). Because anyway we cannot use this targetmanager for AWS.. AWS may need whole new set of configs to pull logs out of it.

gcp-logs ?

Reply:
>gcp-logs
I like it. It also hides implementation details of pubsub. I will make it `gcplog` (consistent with say `syslog`) and tmrw may be `awslog`